src/lib/io/include/sol-uart.h

```diff
@@ -81,22 +90,57 @@ enum sol_uart_parity {
  * @brief Amount of stop bits.
  */
 enum sol_uart_stop_bits {
-    SOL_UART_STOP_BITS_ONE = 0,
-    SOL_UART_STOP_BITS_TWO
+    SOL_UART_STOP_BITS_ONE = 0, /**< Use one stop bit*/
+    SOL_UART_STOP_BITS_TWO /**< Use two stop bits */
 };
 
+/**
+ * @struct sol_uart_config
+ * @brief A configuration struct used to set the UART paramenters.
+ *
+ * @see sol_uart_open()
+ * @note UART follows the Soletta stream guidelines, which can be found here: @ref streams
+ */
 struct sol_uart_config {
 #ifndef SOL_NO_API_VERSION
-#define SOL_UART_CONFIG_API_VERSION (1)
-    uint16_t api_version;
+#define SOL_UART_CONFIG_API_VERSION (1) /**< compile time API version to be checked during runtime */
+    uint16_t api_version; /**< must match #SOL_UART_CONFIG_API_VERSION at runtime */
 #endif
-    enum sol_uart_baud_rate baud_rate;
-    enum sol_uart_data_bits data_bits;
-    enum sol_uart_parity parity;
-    enum sol_uart_stop_bits stop_bits;
-    void (*rx_cb)(void *user_data, struct sol_uart *uart, uint8_t byte_read); /** Set a callback to be called every time a character is received on UART */
-    const void *rx_cb_user_data;
-    bool flow_control; /** Enables software flow control(XOFF and XON) */
+    /**
+     * @brief Callback containing data was read from UART
+     *
+     * @param data User data as provided in sol_uart_config::user_data
+     * @param uart The UART handle
+     * @param buf The UART data
+     *
+     * @return The number of bytes read from @c buf or negative errno on error.
+     */
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, const struct sol_buffer *buf);
+    /**
+     * @brief Informs that a write operation has ended.
+     *
+     * @param data User data as provided in sol_uart_config::user_data
+     * @param uart The UART handle
+     * @param blob The blob that was written
+     * @param status 0 on success or negative errno on error
+     * @note There is no need to call sol_blob_unref().
+     * @see sol_uart_write()
+     */
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    const void *user_data; /**< User data to sol_uart_config::tx_cb() and  sol_uart_config::rx_cb() */
+    /**
+     * @brief The tx buffer max size. The value @c 0 means unlimited data.
+     */
+    size_t tx_size;
+    /**
+     * @brief The rx buffer max size. The value @c 0 means to use the default size. On Linux the default is 4096 and on Riot is 128
```

Comment:
actually, the upper limit may be misleading, some libraries may round that value to offer better performance and the number may be slightly different. We can write this and say the actual available amount is `buf->used`